### PR TITLE
[EVENT] add log reporter

### PR DESCRIPTION
**Pull request by Basasuya**
This PR is based on the previous PR: https://github.com/ray-project/ray/pull/9657
Ray Event Design Doc: https://docs.google.com/document/d/1QSJ1UUEv6zBh6UjVY24w3kkpNdZq4hOGdU7TXgs8YrE/edit#heading=h.ej1vw0pzfkg0

I add the log reporter to support write event to the files

every process will write the events to the file (event_{SOURCE_TYPE}.txt) just like below:
```json
{"time_stamp":"2020-08-29 14:18:15.998084","severity":"INFO","label":"label 1","event_id":"de150792ceb151c815d359d4b675fcc6266a","source_type":"CORE_WORKER","host_name":"Macbool.local","pid":"20830","message":"send message 1","custom_fields":{"task_id":"task 1","job_id":"job 1","node_id":"node 1"}}
```
## Why are these changes needed?

Ray needs an event framework to report critical information, this information includes application-level data like job-condition, node adds/remove, alert, error, and so on. event messages can be sent to different destinations such as the dashboard server, the file. Users can see this information in real-time, or analyze the historical information after the program ends.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/protobuf/event.proto [hunk 1]
```diff
@@ -35,8 +35,8 @@ message Event {
   string label = 6;
   // message is main information of this event
   string message = 7;
-  // timestamp is the report millionseconds since 00:00, Jan 1 1970 UTC
-  int64 timestamp = 8;
+  // timestamp is the format string like 2020-08-29 14:01:02.016101
+  string timestamp = 8;
```
Comment (ashione) [inline src/ray/protobuf/event.proto:39]: Can you explain that why string format is better than int64?

File: src/ray/util/event.cc [hunk 2]
```diff
@@ -16,11 +16,215 @@
...
+    file_name_ += "_" + std::to_string(getpid());
+  }
+
+  log_sink_ = std::unique_ptr<LogFileObject>(new LogFileObject(log_dir_, file_name_));
+}
+
+std::string LogEventReporter::EventToString(const rpc::Event &event) {
+  std::stringstream result;
+
+  result << "{\"time_stamp\":\"" << event.timestamp() << "\",\"severity\":\""
```
Comment (ashione) [inline src/ray/util/event.cc:162]: boost::property_tree::json_parser::write_json could be better if we use it here

File: src/ray/util/event.cc [hunk 3]
```diff
@@ -16,11 +16,215 @@
...
+
+  if (force_flush || bytes_since_flush_ > max_flush_size_) {
+    FlushUnlocked();
+  }
+}
+
+///
+/// LogEventReporter
+///
+LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type, std::string log_dir)
```
Comment (ashione) [inline src/ray/util/event.cc:143]: LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type, const std::string &log_dir)

File: src/ray/util/event.cc [hunk 4]
```diff
@@ -16,11 +16,215 @@
...
+  }
+}
+
+void LogFileObject::Write(const char *message, int message_len, bool force_flush) {
+  absl::MutexLock lock(&mutex_);
+  if ((file_length_ >> 20) >= max_log_size_) {
+    if (file_ != NULL) {
+      fclose(file_);
+      // Rename file when the file size is excess max_log_size_
+      rotate_index_++;
```
Comment (ashione) [inline src/ray/util/event.cc:99]: How to store rotate_index if worker crashed and then reconstructed again?
BTW,  so many low level api (FILE*/fclose/flush) have been used here. I'm not sure whether it can be transplanted to WINDOWS.

File: src/ray/util/event.cc [hunk 5]
```diff
@@ -16,11 +16,215 @@
...
+
+void LogFileObject::FlushLocked() {
+  absl::MutexLock lock(&mutex_);
+  FlushUnlocked();
+}
+
+void LogFileObject::LoopFlush() {
+  if (!shutdown_) {
+    FlushLocked();
+    if (!shutdown_) {
```
Comment (ashione) [inline src/ray/util/event.cc:66]: Leave some comments about double check between Flushblocked.

File: src/ray/util/event.cc [hunk 6]
```diff
@@ -16,11 +16,215 @@
...
+
+std::unique_ptr<EventManager> EventManager::instance_ = nullptr;
+absl::Mutex EventManager::mutex_;
+
 EventManager &EventManager::Instance() {
-  static EventManager instance_;
-  return instance_;
+  if (instance_ == nullptr) {
+    absl::MutexLock lock(&mutex_);
+    if (instance_ == nullptr) {
```
Comment (ashione) [inline src/ray/util/event.cc:213]: It seems duplicated null-check?

File: src/ray/util/event_test.cc [hunk 7]
```diff
@@ -55,14 +55,110 @@ void CheckEventDetail(rpc::Event &event, std::string job_id, std::string node_id
...
+    std::for_each(threads.begin(), threads.end(), [](std::thread &x) { x.join(); });
+  } else {
+    event_context_init();
+    for (int loop_i = 0; loop_i < loop_times; loop_i++) {
+      loop_function(loop_i);
+    }
+  }
+}
+
+std::vector<std::string> ReadEventFromFile(std::string log_file) {
```
Comment (ashione) [inline src/ray/util/event_test.cc:140]: It's not suggested to return a right value vector.

File: src/ray/util/event.cc (hunk 4 above)
Comment (Basasuya) [reply src/ray/util/event.cc:99]: First, if the worker reconstructed, log reporter will first detect all the log file in the event directory, then it will get the new rotate_index
```C++
rotate_index_ = 0;
  for (int i = 1; i <= max_rotate_num_; ++i) {
    if (access(GetFileNameByIndex(i).c_str(), 0) != -1) {
      rotate_index_++;
    }
  }
```

Second, it can not be transplanted to WINDOWS, I think. But if we just use the Glog, it‘s hard to define some complicated functions such as rotate, timing flush, and so on. Do you have some good idea(In Ant, We are going to use boost::log at the beginning, but there are compilation errors in some low versions of GCC)

File: src/ray/util/event.cc (hunk 6 above)
Comment (Basasuya) [reply src/ray/util/event.cc:213]: Here to prevent multi-thread calling Instance API. I have to guarantee the event manager is constructed only once.

File: src/ray/util/event.cc (hunk 4 above)
Comment (ashione) [reply src/ray/util/event.cc:99]: @Basasuya How to generate new rotate index? I mean you may not find the last rotate index if all suffix file names had been created. For example, max_rotate_num is 2, then both xx_1.log and xx_2.log will be in this directory.

For second issue, I test spdlog and it looks good for generating any format logging or something like that.

File: src/ray/util/event.cc (hunk 6 above)
Comment (ashione) [reply src/ray/util/event.cc:213]: So why not add lock_gurad outside

File: src/ray/protobuf/event.proto (hunk 1 above)
Comment (Basasuya) [reply src/ray/protobuf/event.proto:39]: Maybe string format is readable than int64? User doesn't have to transfer to a readable time structure

File: src/ray/util/event.cc (hunk 2 above)
Comment (Basasuya) [reply src/ray/util/event.cc:162]: I will try to use this.

File: src/ray/util/event.cc (hunk 6 above)
Comment (Basasuya) [reply src/ray/util/event.cc:213]: Obviously it is more efficient. If the instance is already initialized, we don't need to add lock.

File: src/ray/util/event_test.cc (hunk 7 above)
Comment (Basasuya) [reply src/ray/util/event_test.cc:140]: Do you mean use 
```C++
const std::vector<std::string>& vc = ReadEventFromFile(log_dir + "/event_RAYLET.log");
```
Or modify the function to 
```C++
ReadEventFromFile(vector<string>& vc, string filename)
```

File: src/ray/protobuf/event.proto (hunk 1 above)
Comment (rkooo567) [reply src/ray/protobuf/event.proto:39]: I think the users of event framework (e.g., dashboard) can just take care of parsing timestamp (and we can just provide timestamp). Let's just use int64 timestamp (for generality)

**Comment by rkooo567** [comment]
Why not just using glog for writing framework? What's the reason why we'd like to develop our own file writing framework?

File: src/ray/util/event_test.cc (hunk 7 above)
Comment (ashione) [reply src/ray/util/event_test.cc:145]: > Do you mean use
> 
> ```c++
> const std::vector<std::string>& vc = ReadEventFromFile(log_dir + "/event_RAYLET.log");
> ```
> 
> Or modify the function to
> 
> ```c++
> ReadEventFromFile(vector<string>& vc, string filename)
> ```

We prefer to second definition. But you need comment what input parameters and output parameters.

File: src/ray/util/event.cc (hunk 6 above)
Comment (ashione) [reply src/ray/util/event.cc:211]: It looks so weird. You'd better make sure it's lock-free if we should use double-check here.

File: src/ray/util/event.cc [hunk 8]
```diff
@@ -16,11 +16,213 @@
...
+  }
+  return *instance_;
+}
+
+EventManager::~EventManager() {
+  for (const auto &element : reporter_map_) {
+    (element.second)->Close();
+  }
+  reporter_map_.clear();
+  io_work_.~work();
```
Comment (ashione) [inline src/ray/util/event.cc:223]: Why call deconstructor here? It will be executed before end of EventManager deconstructor.

File: src/ray/util/event.cc [hunk 9]
```diff
@@ -16,11 +16,213 @@
...
-  static EventManager instance_;
-  return instance_;
+  if (instance_ == nullptr) {
+    absl::MutexLock lock(&mutex_);
+    if (instance_ == nullptr) {
+      instance_ = std::unique_ptr<EventManager>(new EventManager());
+    }
+  }
+  return *instance_;
+}
```
Comment (ashione) [inline src/ray/util/event.cc:216]:
```cpp
EventManager &EventManager::Instance() {
   std::unique_ptr<EventManager> EventManager::instance_(new EventManager());
  return *instance_;
}
```

File: src/ray/protobuf/event.proto (hunk 1 above)
Comment (ashione) [reply src/ray/protobuf/event.proto:39]: @Basasuya You may define two parameters including both string timestamp and timezone if we keep it. But in int64 time format, any component can convert it to other timezone.

**Comment by ashione** [comment]
> Why not just using glog for writing framework? What's the reason why we'd like to develop our own file writing framework?


First, I aslo disagree with repeated development of file systems in ray. But glog has low customizability, so we have to face these:
* No rotating (we need automatically clean log files genreated by long running jobs),
* Can be initialized only once,
* No custom format pattern.

I'd like to replace glog with spdlog https://github.com/gabime/spdlog/tree/v1.x/include/spdlog (popular and good performance).
Or making spdlog as an option.

File: src/ray/protobuf/event.proto (hunk 1 above)
Comment (Basasuya) [reply src/ray/protobuf/event.proto:39]: OK, I will change to the original defination.

File: src/ray/util/event.cc (hunk 4 above)
Comment (Basasuya) [reply src/ray/util/event.cc]: I agree with your opinion, I have changed to the spdlog

**Comment by ashione** [comment]
@Basasuya LGTM. Could you check the rest of comments and reslove them.

**Comment by Basasuya** [comment]
> @Basasuya LGTM. Could you check the rest of comments and reslove them.

Fi

File: src/ray/util/event.cc [hunk 10]
```diff
@@ -15,6 +15,100 @@
...
+  const std::string &message = event.message();
+  for (int i = 0, len = message.size(); i < len; ++i) {
+    if (message[i] == '\n' || message[i] == '\r') {
+      message_buffer << "\\n";
+    } else if (message[i] == '\"') {
+      message_buffer << "\'";
+    } else {
+      message_buffer << message[i];
+    }
+  }
```
Comment (ashione) [inline src/ray/util/event.cc:83]: Is same function with 1. replace '\n' or '\r' to '\\n' 2. put message to buffer together?

File: src/ray/util/event.cc [hunk 11]
```diff
@@ -15,6 +15,100 @@
...
+  boost::property_tree::json_parser::write_json(ss, pt, false);
+
+  return ss.str();
+}
+
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event);
+  result.pop_back();  // pop '\n'
```
Comment (ashione) [inline src/ray/util/event.cc:104]: You'd better make sure the comment is full sentence starts with uppercase and ends with `.` at front of code.

File: src/ray/util/event.h [hunk 12]
```diff
@@ -42,6 +49,36 @@ class BaseEventReporter {
...
+  virtual void Report(const rpc::Event &event) override;
+
+  virtual void Close() override {}
+
+  virtual void Flush();
+
+  virtual std::string GetReporterKey() override { return "glog.event.reporter"; }
+
+ protected:
+  const int rotate_max_file_size_ = 100;  // MB
```
Comment (ashione) [inline src/ray/util/event.h:73]: Why these rotate fields can not be configurable?

File: src/ray/util/event.cc (hunk 10 above)
Comment (Basasuya) [reply src/ray/util/event.cc]: I have removed this part of the code, because I find boost JSON support escape character transfer.

**Review by ashione — APPROVED** [review]

**Comment by ashione** [comment]
@rkooo567 @raulchen @yuyiming Could you merge it if travis is green.

**Comment by ashione** [comment]
@Basasuya More comments should be better for more readable and advance development before all public functions.

**Comment by rkooo567** [comment]
I will review this very soon. Have been too busy by Ray 1.0 related stuff.

File: src/ray/util/event.cc [hunk 13]
```diff
@@ -15,6 +15,93 @@
...
+                                   int rotate_max_file_size, int rotate_max_file_num)
+    : log_dir_(log_dir),
+      force_flush_(force_flush),
+      rotate_max_file_size_(rotate_max_file_size),
+      rotate_max_file_num_(rotate_max_file_num) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
```
Comment (rkooo567) [inline src/ray/util/event.cc:32]: Let's remove boost::filesystem from dependencies and share our log directory here. I don't think there's any situation we'd like to have a separate directory only for events.

File: src/ray/util/event.cc [hunk 14]
```diff
@@ -15,6 +15,93 @@
...
+    file_name_ += "_" + std::to_string(getpid());
+  }
+  file_name_ += ".log";
+
+  std::string log_sink_key = GetReporterKey() + log_dir_ + file_name_;
+  log_sink_ = spdlog::get(log_sink_key);
+  if (log_sink_ == nullptr) {
+    log_sink_ =
+        spdlog::rotating_logger_mt(log_sink_key, log_dir_ + file_name_,
+                                   1048576 * rotate_max_file_size_, rotate_max_file_num_);
```
Comment (rkooo567) [inline src/ray/util/event.cc:45]: Can you comment the meaning of this number?

File: src/ray/util/event.cc [hunk 15]
```diff
@@ -15,6 +15,93 @@
...
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event);
+  // Pop the last character from the result string because it is breakline '\n'.
+  result.pop_back();
+
+  log_sink_->info(result);
+  if (force_flush_) {
+    Flush();
```
Comment (rkooo567) [inline src/ray/util/event.cc:101]: When do we need this attribute? Is it for testing? Can you clearly specify this attribute is for testing?

File: src/ray/util/event.cc [hunk 16]
```diff
@@ -15,6 +15,93 @@
...
+  boost::property_tree::json_parser::write_json(ss, pt, false);
+
+  return ss.str();
+}
+
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event);
+  // Pop the last character from the result string because it is breakline '\n'.
```
Comment (rkooo567) [inline src/ray/util/event.cc:96]: Will logs still have newline without this?

File: src/ray/util/event_test.cc [hunk 17]
```diff
@@ -107,6 +204,112 @@ TEST(EVENT_TEST, TEST_BASIC) {
   CheckEventDetail(result[3], "", "", "", "GCS", "FATAL", "", "");
 }
 
+TEST(EVENT_TEST, LOG_ONE_THREAD) {
+  std::string log_dir = GenerateLogDir();
```
Comment (rkooo567) [inline src/ray/util/event_test.cc:208]: I prefer not to have actual filesystem operation in uni tests. is there any way we can mock this out?

File: src/ray/util/event.cc (hunk 13 above)
Comment (Basasuya) [reply src/ray/util/event.cc:32]: So your opinion is that we don't need to create the log folder. This operation should be placed outside the framework?
When the log reporter detected this log folder is not created, It means that the log reporter would stop working.

I have a question about which framework would responsible for creating the log folder(Or just the same as the RAY_LOG log_dir)

File: src/ray/util/event.cc (hunk 15 above)
Comment (Basasuya) [reply src/ray/util/event.cc:101]: force_flush_ is used to clear the file buffer. When fwrite function is called, sometimes the information is not actually written to the file, but stored in the OS file buffer. The flush will be triggered when it exceeds certain buffer size.

File: src/ray/util/event.cc (hunk 13 above)
Comment (rkooo567) [reply src/ray/util/event.cc:32]: What I am saying is we can just write event logs to the same directory we record log files /tmp/ray/session_*, where RAY_LOG writes log files. I just couldn’t think of any scenario these logs should be recorded in any separate dir from there.

File: src/ray/util/event.cc (hunk 15 above)
Comment (rkooo567) [reply src/ray/util/event.cc:101]: I understand when we need flush! But isn’t this triggered automatically by spdlog if buffers are filled up? I just don’t know when we need to set this attribute true except test scenario.

File: src/ray/util/event.cc (hunk 16 above)
Comment (Basasuya) [reply src/ray/util/event.cc:96]: Because boost::property_tree::json_parser::write_json would generate'\n' at the end of string.
Then. every string which the spdlog write would also generate a newline.
So I should remove the '\n' at the end of the string.

File: src/ray/util/event_test.cc (hunk 17 above)
Comment (Basasuya) [reply src/ray/util/event_test.cc:208]: This uni test is designed to test the logic of the log reporter. Concretely speaking, I will test the correctness of the interface in single thread and multi-thread and the correctness of rotating file.At least for now, I think it’s hard to mock this function.

File: src/ray/util/event.cc (hunk 15 above)
Comment (Basasuya) [reply src/ray/util/event.cc:101]: When the system receives SIGTERM, the log reporter's destructor may not be triggered. Some of the last event information may not be sent successfully. But if it is force_flush mode, because it will be flushed when RAY_EVENT is called, so we don’t worry about this problem.

File: src/ray/util/event.cc (hunk 13 above)
Comment (Basasuya) [reply src/ray/util/event.cc:32]: OK, I would remove the boost filesystem framework.

File: src/ray/util/event_test.cc (hunk 17 above)
Comment (ashione) [reply src/ray/util/event_test.cc:209]: @rkooo567 Could you tell us what's your main concern in testing log files rotation? If we have to check whether tiny event log files can be rotated in our test cases.

**Review by ericl — CHANGES_REQUESTED** [review]
Hey guys, do we really need spdlog here? I would much rather keep glog, even if it has slightly less features, rather than add a new dependency on some random library.

**Comment by ashione** [comment]
> Hey guys, do we really need spdlog here? I would much rather keep glog, even if it has slightly less features, rather than add a new dependency on some random library.

Hi eric, my answer is yes. We do need spdlog becasue of its features and performance.

In our production business and best practices, we not only need to support a large number of long-running jobs, some are even online jobs. These long-running tasks are usually run in a docker environment, and their specifications are destined to not carry large-capacity disks. Therefore, we must clean up the log files regularly to prevent our long-running tasks from being affected.
Besides, it is not that we have to replace glog, because glog is currently unable to meet business needs.
Let me briefly summarize again:
1. Cannot automatically clean up and rotating the log.
2. There are two duplicate outputs, one is standard output and the other is log file.
3. For each process, glog can only be initialized once. If not, multiple errors are reported and the process exits abnormally. This means that if the user uses cpp worker, the log generated by the user's actor logic will be printed in the same log file as the core worker. Event sender can not create a new file for genrating event stuff as well.
4. The log level cannot be dynamically modified at runtime, which is very helpful for debugging long-running tasks.
5. Unable to customize the log input format, which is fatal for events.

We have also discussed for a long time whether to introduce this dependency. The conclusion is that we firstly introduce spdlog into the event system so that its needs can be met, instead of modifying glog or repeating  filesystem operations.
The next step is to add a switch to select the logging tool. So we can freely choose whether to use glog or spdlog. The design of the RAY_LOG macro allows us to realize these two kinds of compatibility easily.

In addition, it is worth mentioning that the maintenance of glog is indeed inferior to spdlog. My understanding is that adapting to the needs to make some changes will make ray a greater open source software. Instead of avoiding it, endure it temporarily.

cc @Basasuya @raulchen

**Comment by ashione** [comment]
> > Hey guys, do we really need spdlog here? I would much rather keep glog, even if it has slightly less features, rather than add a new dependency on some random library.
> 
> Hi eric, my answer is yes. We do need spdlog becasue of its features and performance.
> 
> In our production business and best practices, we not only need to support a large number of long-running jobs, some are even online jobs. These long-running tasks are usually run in a docker environment, and their specifications are destined to not carry large-capacity disks. Therefore, we must clean up the log files regularly to prevent our long-running tasks from being affected.
> Besides, it is not that we have to replace glog, because glog is currently unable to meet business needs.
> Let me briefly summarize again:
> 
> 1. Cannot automatically clean up and rotating the log.
> 2. There are two duplicate outputs, one is standard output and the other is log file.
> 3. For each process, glog can only be initialized once. If not, multiple errors are reported and the process exits abnormally. This means that if the user uses cpp worker, the log generated by the user's actor logic will be printed in the same log file as the core worker. Event sender can not create a new file for genrating event stuff as well.
> 4. The log level cannot be dynamically modified at runtime, which is very helpful for debugging long-running tasks.
> 5. Unable to customize the log input format, which is fatal for events.
> 
> We have also discussed for a long time whether to introduce this dependency. The conclusion is that we firstly introduce spdlog into the event system so that its needs can be met, instead of modifying glog or repeating filesystem operations.
> The next step is to add a switch to select the logging tool. So we can freely choose whether to use glog or spdlog. The design of the RAY_LOG macro allows us to realize these two kinds of compatibility easily.
> 
> In addition, it is worth mentioning that the maintenance of glog is indeed inferior to spdlog. My understanding is that adapting to the needs to make some changes will make ray a greater open source software. Instead of avoiding it, endure it temporarily.
> 
> cc @Basasuya @raulchen

**Comment by Basasuya** [comment]
@ericl @rkooo567 @ashione We have discussed this in slack 
https://ray-distributed.slack.com/archives/C016MPZE1L1/p1598937145003600

**Comment by ericl** [comment]
https://stackoverflow.com/questions/16439857/google-glog-rolling-files seems to indicate glog supports log rotation, have you looked into this?

**Comment by rkooo567** [comment]
@ericl I think the hardest blocker for using glog is 3 (imo, other features are not "absolutely needed" for OSS). For events, we'd like to write them in a separate file from regular logs. 

It seems like there's a way to implement additional log sink for glog, but it looked a little sketchy (https://ray-distributed.slack.com/archives/C016MPZE1L1/p1598943141006800?thread_ts=1598937145.003600&cid=C016MPZE1L1).

**Comment by ericl** [comment]
Hmm I can understand the desire to write to a separate file, but why not instead just write at DEBUG or INFO level into the normal logs?. It is useful to correlate these events with other log message, and easy enough to filter with grep if needed. Is the concern the volume of logs is way too high?

**Comment by ashione** [comment]
> Hmm I can understand the desire to write to a separate file, but why not instead just write at DEBUG or INFO level into the normal logs?. It is useful to correlate these events with other log message, and easy enough to filter with grep if needed. Is the concern the volume of logs is way too high?

It is better not to mix the information of event with the normal log, otherwise this is not the so-called event, but a special log item. The purpose of putting the event information into a separate file is to facilitate the distributed retrieval or index building of dashboards or other processes. And you said that using grep to find is our way of positioning artificially.

**Comment by ashione** [comment]
> https://stackoverflow.com/questions/16439857/google-glog-rolling-files seems to indicate glog supports log rotation, have you looked into this?

The current version of GLOG does not support rotation because it produces a random file with the current timestamp as the suffix. On GLOG's GitHub homepage, I observed that there was a PR about rotation a few years ago, and it has not been merged. Internally, we modified GLOG to achieve it.

**Comment by ashione** [comment]
> @ericl I think the hardest blocker for using glog is 3 (imo, other features are not "absolutely needed" for OSS). For events, we'd like to write them in a separate file from regular logs.
> 
> It seems like there's a way to implement additional log sink for glog, but it looked a little sketchy (https://ray-distributed.slack.com/archives/C016MPZE1L1/p1598943141006800?thread_ts=1598937145.003600&cid=C016MPZE1L1).

This way of adding custom sink also requires us to manage files and decide when and whether rotation is needed.

**Comment by Basasuya** [comment]
@ericl I don't think we should use glog at DEBUG or INFO level to send event messages. The first reason is that glog does not support multiple logger object of one thread. In this case, RAY_EVENT and RAY_LOG will share glog object, then glog can not write to different files. Second,  If we redirected DEBUG, FATAL mode to write events files, these two severities are not enough in our practical experience. At the same time, it will affect the function of  RAY_LOG.
In summary, glog doesn't support multiple logger object of one thread natively. And glog can not rotate file just like other log frameworks such as log4j, spdlog. That's the reason why we want to try another framework.

**Comment by ericl** [comment]
Ok, the lack of log rotation does seem like a deal breaker. Frankly, saving
events at debug level does seem fine though. There is no principled reason
why it can't be done that way.

I'm fine with adding this dependency due to the lack of rotation, but in
the future we should really try as hard as we can to avoid depending on
external libraries.

**Comment by Basasuya** [comment]
@rkooo567 Could you please review my reply to your previous comments?
https://github.com/ray-project/ray/pull/10419#discussion_r486277040
https://github.com/ray-project/ray/pull/10419#discussion_r486272075
https://github.com/ray-project/ray/pull/10419#discussion_r486757571

**Review by ericl — APPROVED** [review]
